exchanges/okx/okx_websocket.go

```diff
@@ -0,0 +1,1844 @@
+		if err != nil {
+			return err
+		}
+		var baseVolume float64
+		var quoteVolume float64
+		if a == asset.Spot || a == asset.Margin {
+			baseVolume = response.Data[i].Vol24H
+			quoteVolume = response.Data[i].VolCcy24H
+		} else {
+			baseVolume = response.Data[i].VolCcy24H
+			quoteVolume = response.Data[i].Vol24H
+		}
+		ok.Websocket.DataHandler <- &ticker.Price{
+			ExchangeName: ok.Name,
+			Open:         response.Data[i].Open24H,
+			Volume:       baseVolume,
+			QuoteVolume:  quoteVolume,
+			High:         response.Data[i].High24H,
+			Low:          response.Data[i].Low24H,
+			Bid:          response.Data[i].BidPrice,
+			Ask:          response.Data[i].BestAskPrice,
+			BidSize:      response.Data[i].BidSize,
+			AskSize:      response.Data[i].BestAskSize,
+			Last:         response.Data[i].LastTradePrice,
+			AssetType:    a,
+			Pair:         c,
+			LastUpdated:  response.Data[i].TickerDataGenerationTime,
+		}
+	}
+	return nil
+}
+
+// GenerateDefaultSubscriptions returns a list of default subscription message.
+func (ok *Okx) GenerateDefaultSubscriptions() ([]stream.ChannelSubscription, error) {
+	var subscriptions []stream.ChannelSubscription
+	assets := ok.GetAssetTypes(true)
+	if ok.Websocket.CanUseAuthenticatedEndpoints() {
+		defaultSubscribedChannels = append(defaultSubscribedChannels,
+			okxChannelAccount,
+			okxChannelOrders,
+		)
+	}
+	for x := range assets {
+		pairs, err := ok.GetEnabledPairs(assets[x])
+		if err != nil {
+			return nil, err
+		}
+		for y := range defaultSubscribedChannels {
+			if defaultSubscribedChannels[y] == okxChannelCandle5m ||
+				defaultSubscribedChannels[y] == okxChannelTickers ||
+				defaultSubscribedChannels[y] == okxChannelOrders ||
+				defaultSubscribedChannels[y] == okxChannelOrderBooks ||
+				defaultSubscribedChannels[y] == okxChannelOrderBooks5 ||
+				defaultSubscribedChannels[y] == okxChannelOrderBooks50TBT ||
+				defaultSubscribedChannels[y] == okxChannelOrderBooksTBT ||
+				defaultSubscribedChannels[y] == okxChannelTrades {
+				for p := range pairs {
+					subscriptions = append(subscriptions, stream.ChannelSubscription{
+						Channel:  defaultSubscribedChannels[y],
+						Asset:    assets[x],
+						Currency: pairs[p],
+					})
+				}
+			} else {
+				subscriptions = append(subscriptions, stream.ChannelSubscription{
+					Channel: defaultSubscribedChannels[y],
+				})
+			}
+		}
+	}
+	return subscriptions, nil
+}
+
+// wsProcessPushData processes push data coming through the websocket channel
+func (ok *Okx) wsProcessPushData(data []byte, resp interface{}) error {
+	if err := json.Unmarshal(data, resp); err != nil {
+		return err
+	}
+	ok.Websocket.DataHandler <- resp
+	return nil
+}
+
+// Websocket Trade methods
+
+// WSPlaceOrder places an order thought the websocket connection stream, and returns a SubmitResponse and error message.
+func (ok *Okx) WSPlaceOrder(arg *PlaceOrderRequestParam) (*PlaceOrderResponse, error) {
```

Comment:
I think these functions (including `WsAmendMultipleOrders`, `WsPlaceMultipleOrder`, `WsCancelOrder`, `WsCancleMultipleOrder`, `WsAmendOrder`) should have tests written for them. If you would like help in setting up test functions to connect to the websocket, let me know. For a quick reference, zb_test.go has the function `func setupWsAuth(t *testing.T) {` to help setup an auth ws connection and use the functions